toolchain/check/cpp/import.cpp

```diff
@@ -1228,25 +1235,27 @@ static auto MakeOptionalType(Context& context, SemIR::LocId loc_id,
   return ExprAsType(context, loc_id, call_id);
 }
 
-// Maps a C++ pointer type to a Carbon pointer type.
+// Maps a C++ pointer type to a Carbon pointer type. If the pointer is wrapped
+// with a qualified type, maps it as well to keep qualifiers on the pointer.
+// We need to do it here because we wrap nullable pointers with an optional type
+// and the qualifiers need to be set to the pointer and not the optional type.
 static auto MapPointerType(Context& context, SemIR::LocId loc_id,
-                           clang::QualType type, TypeExpr pointee_type_expr)
-    -> TypeExpr {
+                           clang::QualType type, clang::QualType qualified_type,
+                           TypeExpr pointee_type_expr) -> TypeExpr {
   CARBON_CHECK(type->isPointerType());
 
-  bool optional = false;
-  if (auto nullability = type->getNullability();
-      !nullability.has_value() ||
-      *nullability != clang::NullabilityKind::NonNull) {
-    // If the type was produced by C++ template substitution, then we assume it
-    // was deduced from a Carbon pointer type, so it's non-null.
-    if (!type->getAs<clang::SubstTemplateTypeParmType>()) {
-      optional = true;
-    }
-  }
+  bool optional =
+      !IsClangTypeNonNull(type) &&
+      // If the type was produced by C++ template substitution, then we assume
+      // it was deduced from a Carbon pointer type, so it's non-null.
+      !type->getAs<clang::SubstTemplateTypeParmType>();
 
   TypeExpr pointer_type_expr = TypeExpr::ForUnsugared(
       context, GetPointerType(context, pointee_type_expr.inst_id));
+  if (!qualified_type.isNull()) {
+    pointer_type_expr =
+        MapQualifiedType(context, qualified_type, pointer_type_expr);
```

Comment:
It kooks like the intent here is to map `T* const` to `Optional(const (T*))` instead of `const Optional(T*)`. (Or am I misunderstanding?) I think that's incorrect -- the latter type would allow us to assign to the optional, which shouldn't be allowed for a const C++ object such as a `T* const`.

Reply:
FWIW, this specific error in the tests we have now no longer triggers because we remove the `const` qualifier from the pointer param.
I don't think it changes the conclusions here though.